examples/author_disambiguation_advanced.py

```diff
@@ -0,0 +1,366 @@
+        ]))])
+
+    # Train a classifier on these vectors
+    classifier = GradientBoostingClassifier(n_estimators=500,
+                                            max_depth=9,
+                                            max_features=10,
+                                            learning_rate=0.125,
+                                            verbose=3)
+
+    # Return the whole pipeline
+    estimator = Pipeline([("transformer", transformer),
+                          ("classifier", classifier)]).fit(X, y)
+
+    return estimator
+
+
+def affinity(X, estimator, step=10000):
+    """Custom affinity function, using a pre-learned distance estimator."""
+    all_i, all_j = np.triu_indices(len(X), k=1)
+    n_pairs = len(all_i)
+    distances = np.zeros(n_pairs)
+
+    for start in range(0, n_pairs, step):
+        end = min(n_pairs, start+step)
+        Xt = np.empty((end-start, 2), dtype=np.object)
+
+        for k, (i, j) in enumerate(zip(all_i[start:end],
+                                       all_j[start:end])):
+            Xt[k, 0], Xt[k, 1] = X[i, 0], X[j, 0]
+
+        Xt = estimator.predict_proba(Xt)[:, 1]
+        distances[start:end] = Xt[:]
+
+    return squareform(distances)
+
+
+def blocking(X):
+    """Blocking function using last name and first initial as key."""
+    def last_name_first_initial(name):
+        names = name.split(",", 1)
+
+        if len(names) == 2:
+            name = "%s %s" % (names[0], names[1].strip()[0])
+        else:
+            name = names[0]
+
+        name = normalize_name(name)
+        return name
+
+    blocks = []
+
+    for signature in X[:, 0]:
+        blocks.append(last_name_first_initial(signature["author_name"]))
+
+    return np.array(blocks)
+
+
+if __name__ == "__main__":
+    # Load paired data
+    X, y, signatures, records = pickle.load(open(sys.argv[1], "r"))
+    signatures, records = resolve_publications(signatures, records)
+
+    Xt = np.empty((len(X), 2), dtype=np.object)
+    for k, (i, j) in enumerate(X):
+        Xt[k, 0] = signatures[i]
+        Xt[k, 1] = signatures[j]
+    X = Xt
+
+    # Learn a distance estimator on paired signatures
+    distance_estimator = build_distance_estimator(X, y)
+
+    # Load signatures to cluster
+    signatures, truth, records = pickle.load(open(sys.argv[2], "r"))
+    _, records = resolve_publications(signatures, records)
+
+    X = np.empty((len(signatures), 1), dtype=np.object)
+    for i, signature in enumerate(signatures):
+        X[i, 0] = signature
+
+    # Semi-supervised block clustering
+    train, test = train_test_split(np.arange(len(X)),
+                                   test_size=0.9, random_state=42)
+    y = -np.ones(len(X), dtype=np.int)
+    y[train] = truth[train]
+
+    clusterer = BlockClustering(
+        blocking=blocking,
+        base_estimator=ScipyHierarchicalClustering(
+            threshold=0.9995,
+            affinity=partial(affinity, estimator=distance_estimator),
```

Comment:
This seems to be a bottleneck, as the learned distance estimator needs to be pickled for every job launched by the clusterer. A workaround is put the estimator variable in global, hence avoiding the transfer. I'll investigate whether there is a cleaner workaround than this ugly hack...